tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -161,15 +164,15 @@ public static void fastCompleteSpan(Map<String, String> metadata) {
      * Like {@link #completeSpan()}, but adds {@code metadata} to the current span being completed.
      */
     public static Optional<Span> completeSpan(Map<String, String> metadata) {
+        boolean observable = isTraceObservable();
+        Trace trace = currentTrace.get();
         Optional<Span> maybeSpan = popCurrentSpan()
-                .map(openSpan -> toSpan(openSpan, metadata));
+                .map(openSpan -> toSpan(openSpan, metadata, trace.getTraceId()));
 
         // Notify subscribers iff trace is observable
-        maybeSpan.ifPresent(span -> {
-            if (currentTrace.get().isObservable()) {
-                notifyObservers(span);
-            }
-        });
+        if (maybeSpan.isPresent() && observable) {
```

Comment:
should we reuse the trace to avoid additional thread local get and simplify?

```suggestion
        if (isObservable(trace)) {
            maybeSpan.ifPresent(Tracer::notifyObservers);
        }
```

Reply:
👍